nemo_skills/pipeline/utils/declarative.py

```diff
@@ -0,0 +1,703 @@
+
+    # Create pipeline with dependency graph
+    pipeline = Pipeline(
+        name="full_pipeline",
+        cluster_config=cluster_config,
+        jobs=[
+            {"name": "prep", "group": prep_group},
+
+            # Multi-group heterogeneous job (both eval groups in ONE SLURM job with 2 het components)
+            {"name": "evals", "groups": [group_8b, group_32b], "dependencies": ["prep"]},
+
+            # Report depends on the multi-group eval job
+            {"name": "report", "group": report_group, "dependencies": ["evals"]},
+    pipeline.run()
+"""
+
+import logging
+import shlex
+from contextlib import nullcontext
+from dataclasses import dataclass, field
+from typing import Callable, Dict, List, Optional, Tuple, Union
+
+import nemo_run as run
+
+from nemo_skills.pipeline.utils import (
+    get_env_variables,
+    get_executor,
+    get_exp,
+    get_exp_handles,
+    get_tunnel,
+    run_exp,
+    temporary_env_update,
+)
+from nemo_skills.pipeline.utils.commands import wrap_command
+from nemo_skills.pipeline.utils.exp import REUSE_CODE_EXP, get_packaging_job_key, install_packages_wrap, tunnel_hash
+from nemo_skills.pipeline.utils.mounts import is_mounted_filepath
+from nemo_skills.pipeline.utils.packager import get_registered_external_repo
+from nemo_skills.utils import get_logger_name
+
+LOG = logging.getLogger(get_logger_name(__file__))
+
+
+@dataclass
+class Command:
+    """Declarative command for running tasks in containers.
+
+    The command can be either:
+    - A string: evaluated immediately
+    - A callable (lambda): evaluated lazily when the task is prepared
+
+    Lambdas are ONLY needed for cross-component references (hostname_ref, meta_ref).
+    The het_group_index isn't assigned until pipeline execution, so these must be lazy:
+        # Lambda is ESSENTIAL here - server.hostname_ref() and meta_ref() don't exist yet
+        client = Command(command=lambda: f"curl {server.hostname_ref()}:{server.meta_ref('port')}")
+    """
+
+    # Command can be a string or callable (lambda).
+    # Lambdas are primarily used for cross-component references (hostname_ref, meta_ref).
+    command: Union[str, Callable]
+    container: str = "nemo-skills"
+    gpus: Optional[int] = None
+    nodes: int = 1
+    name: str = "command"
+    working_dir: str = "/nemo_run/code"
+    env_vars: Dict[str, str] = field(default_factory=dict)
+    installation_command: Optional[str] = None
+    port: Optional[int] = None  # Can be set from metadata
+    metadata: Dict[str, any] = field(default_factory=dict)  # Stores metadata from command builders
+    het_group_index: Optional[int] = None  # Set per-job by Pipeline (not global)
+
+    def __post_init__(self):
+        # Wrap plain strings with environment setup
+        if isinstance(self.command, str) and (self.env_vars or self.working_dir):
+            self.command = wrap_command(self.command, self.working_dir, self.env_vars)
```

Comment:
Just a stylistic choice, but I will suggest methods like `wrap_command` be co-located with the `Command` data class, since they mostly won't have any other utility.

If they do end up with a utility outside, we can always make them a `@staticmethod`.